File: app/src/main/java/com/iomt/android/DeviceViewModel.java
package com.iomt.android;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;

public class DeviceViewModel extends ViewModel {
    private final MutableLiveData<List<AbsCell>> cellsLiveData = new MutableLiveData<>();

    public DeviceViewModel () {resetCells();}

    public void addCell(final AbsCell deviceCell) {
        final List<AbsCell> oldList = cellsLiveData.getValue();
        if (oldList == null) {
            return;
        }
        final List<AbsCell> tmp = new ArrayList<>(oldList);
        tmp.add(deviceCell);
        cellsLiveData.setValue(tmp);
    }

    public void resetCells() {
        final List<AbsCell> tmp = new ArrayList<>();
        cellsLiveData.setValue(tmp);
    }

    public LiveData<List<AbsCell>> getCellsLiveData() {
        return cellsLiveData;
    }
}

File: app/src/main/java/com/iomt/android/DeviceViewModel.kt
@file:Suppress("MISSING_KDOC_CLASS_ELEMENTS", "MISSING_KDOC_ON_FUNCTION")

package com.iomt.android

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import java.util.*

/**
 * Device View Model
 */
class DeviceViewModel : ViewModel() {
    @Suppress("TYPE_ALIAS")
    private val cellsLiveData: MutableLiveData<List<AbstractCell>> = MutableLiveData()

    /**
     * @param deviceCell
     */
    fun addCell(deviceCell: AbstractCell) {
        val oldList = cellsLiveData.value ?: return
        val tmp: MutableList<AbstractCell> = ArrayList(oldList)
        tmp.add(deviceCell)
        cellsLiveData.value = tmp
    }

    fun resetCells() {
        val tmp: List<AbstractCell> = ArrayList()
        cellsLiveData.value = tmp
    }

    @Suppress("TYPE_ALIAS")
    fun getCellsLiveData(): LiveData<List<AbstractCell>> = cellsLiveData

    init {
        resetCells()
    }
}
